File: leet_code/JS/linked_list/detectCycle.js
function detectCycle(head) {
    if (!head || !head.next || !head.next.next) return null

    let first = head.next;
    let second = head.next.next;


    while (first !== second) {
        first = first.next
        // if `second.next` or `second.next.next` does NOT exist, it's not a cycle.
        if (!second.next || !second.next.next) return null
        second = second.next.next
    }
    first = head

    while (second !== first) {
        second = second.next
        first = first.next
    }

    return first
}

File: leet_code/python/linked_list/detect_cycle.py
# O(n) time
# O(n) time
# O(1) space
class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        slow = head
        fast = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

            if fast == slow:
                slow = head

                while fast != slow:
                    fast = fast.next
                    slow = slow.next

                return slow
        return None

